src/main/scala/internal/Graph.scala

```diff
@@ -0,0 +1,425 @@
+      case (nodeId, deps) => (otherNodeIdMap(nodeId), deps.map(otherNodeOrSourceIdMap))
+    }
+    val newSources = sources ++ newSourceIds
+    val newSinkDependencies = sinkDependencies ++ graph.sinkDependencies.map {
+      case (sinkId, dep) => (otherSinkIdMap(sinkId), otherNodeOrSourceIdMap(dep))
+    }
+    val newGraph = Graph(newSources, newSinkDependencies, newOperators, newDependencies)
+
+    (newGraph, otherSourceIdMap, otherSinkIdMap)
+  }
+
+  /**
+   * Immutably attaches an entire other graph onto this one, splicing Sinks of this graph to Sources of
+   * the graph being added. The unspliced Sink, Source, and Node ids of the other graph may be reassigned
+   * to prevent collisions with existing ids. So, this method makes sure to return mappings of old Sink and
+   * Source ids in the graph being added to their newly assigned Sink and Source ids.
+   *
+   * All Sources and Sinks being spliced are removed from the newly created graph.
+   *
+   * @param graph  The graph to connect to this graph
+   * @param spliceMap A map that specifies how to splice the Sources and Sinks.
+   *                  Keys are Sources in the graph being added, and
+   *                  Values are the Sinks in the existing graph to splice them to.
+   *
+   * @return  A triple containing:
+   *          - The new graph
+   *          - A map of old id to new id for sources in the graph being added
+   *          - A map of old id to new id for sinks in the graph being added
+   */
+  def connectGraph(
+      graph: Graph,
+      spliceMap: Map[SourceId, SinkId]
+    ): (Graph, Map[SourceId, SourceId], Map[SinkId, SinkId]) = {
+    require(spliceMap.keys.forall(source => graph.sources.contains(source)),
+      "Must connect to sources that exist in the other graph")
+    require(spliceMap.values.forall(sink => sinks.contains(sink)),
+      "Must connect to sinks that exist in this graph")
+
+    val (addedGraph, otherSourceIdMap, otherSinkIdMap) = addGraph(graph)
+
+    val connectedGraph = spliceMap.foldLeft(addedGraph) {
+      case (curGraph, (oldOtherSource, sink)) =>
+        val source = otherSourceIdMap(oldOtherSource)
+        val sinkDependency = getSinkDependency(sink)
+        curGraph.replaceDependency(source, sinkDependency)
+          .removeSource(source)
+    }
+
+    val newGraph = spliceMap.values.toSet.foldLeft(connectedGraph) {
+      case (curGraph, sink) => curGraph.removeSink(sink)
+    }
+
+    (newGraph, otherSourceIdMap -- spliceMap.keySet, otherSinkIdMap)
+  }
+
+  /**
+   * Replace a set of nodes and their connections by inserting a different graph in their
+   * place, and specifying how to connect it to the existing graph.
+   *
+   * @param nodesToRemove  The set of nodes to remove
+   * @param replacement  The graph to insert in their place
+   * @param replacementSourceSplice  A specification of how to connect the replacement graph to the existing graph.
```

Comment:
Yeah, the assumption is users never touch any of the graph methods directly, and the only method that will need to use this will have enough contextual knowledge to know what connections they want to make.